### ISSUE TITLE
Add burndown-like charts for milestones

### ISSUE THREAD

**Issue by Joshuaalbert** [nikku/wuffle#18]
Dear @nikku, wonderful initiative. I hope waffle goes open source, but even if it doesn't I'd still be interested to support and possibly contribute to wuffle.

One thing that I'd like to see is a burndown chart for milestones.
Because issues are not assigned estimated time-lengths I suggest a Bayesian approach to estimating time to complete a task.

Basically, 
```
Let T[a, l] be the time to complete an issue
a - indexes assignee
l - indexes labels
```

Consider the contrived case where issues arrive randomly and sequentially and are completed by a black box process dependent on `[a, l]`.

The arrival of issues is then a Poisson process.
Therefore, in the case that issues are solved sequentially, the time to solve an issue is exponentially distributed.

The rate parameter for of the exponential distribution can be estimated by,

```
rate[a,l] = N[a, l] / Sum_i T_i[a,l]
             = 1 / (average of all past issue completion lengths done by assignee a with labels l)
```

From this the probability distribution for the time to complete a task by assignee `a` with label `l` is `P(T[a,l]) = Exponential[rate[a,l]]`.

[The sum of many exponentially variables has an analytic mean and variance](https://en.wikipedia.org/wiki/Exponential_distribution#Sum_of_two_independent_exponential_random_variables).

So it would always be possible to estimate the expected completion time of the milestone. And as new issues are added, so would the estimate change. If you assumed that all assignees work in parallel there is also an analogous solution.

**Comment by nikku** [comment]
Thanks for your interest in this project.

> One thing that I'd like to see is a burndown chart for milestones.

I'm not into the Math behind the things you propose but I do see the potential of milestone statistics. 

Before rushing forward and discussing your proposal and the specific algorithms described in depth, let's take a step back. What transitions would we need to collect to generate meaningful statistics?

**Comment by Joshuaalbert** [comment]
Greetings @nikku. This is a great question. Firstly, I acknowledge that Wuffle has the potential to be a great tool for individual developers, small dev teams, and also tech companies building at scale. As long as they wish to use github as the source of truth in their development.

In terms of meta-data that would need to be collected for burndowns and other useful stats, we would need a time series of the events:

1. a new issue is created: who created, time, title, labels, assignee, milestone, project
2. an issue is closed: who closed, time, message (should include commit tag in proper usage).
3. an issue is commented on/modified: who commented, time

Really only 1. and 2. are needed for burndowns. But burndowns can be further processed to yield all sorts of useful information like efficiency per assignee per label, time estimation bias, confidence intervals for completion, and running predictive scenarios (e.g. what would happen if you added another dev or paired two devs up, or tackled the milestone in two parallel sprints. This is most cool if you also allow a cost per person-day per dev, then you can get graphs of completion confidence intervals vs cost confidence intervals which would make any product owner very happy).

If you also allow 3. then you can infer directional communication networks among developers and infer who helps who, which potentially helps locate synergistic dev-pairs.

I'll note that the ability to automatically parse and modify the issue first post would be useful, e.g. the use a pattern to specify the estimated time to do a task, perhaps ```%EstTime 0.5PD``` means half a person-day.

**Comment by nikku** [comment]
Not going to happen any time soon, without external contribution.
